packages/livekit-rtc/src/rpc.ts

```diff
@@ -0,0 +1,172 @@
+// SPDX-FileCopyrightText: 2024 LiveKit, Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+// TODO: This implementation is only a prototype
+//       The final version will use a protocol types where possible
+export class RpcRequest {
+  id: string;
+  method: string;
+  payload: string;
+  responseTimeoutMs: number;
+
+  constructor({
+    method,
+    payload,
+    responseTimeoutMs,
+    id = crypto.randomUUID(),
+  }: {
+    method: string;
+    payload: string;
+    responseTimeoutMs: number;
+    id?: string;
+  }) {
+    this.id = id;
+    this.method = method;
+    this.payload = payload;
+    this.responseTimeoutMs = responseTimeoutMs;
+  }
+}
+
+export class RpcAck {
+  requestId: string;
+
+  constructor({ requestId }: { requestId: string }) {
+    this.requestId = requestId;
+  }
+}
+
+export class RpcResponse {
+  requestId: string;
+  payload?: string;
+  error?: RpcError;
+
+  constructor({
+    requestId,
+    payload,
+    error,
+  }: {
+    requestId: string;
+    payload?: string;
+    error?: RpcError;
+  }) {
+    this.requestId = requestId;
+    this.payload = payload;
+    this.error = error;
+  }
+}
+
+/**
+ * Specialized error handling for RPC methods.
+ *
+ * Instances of this type, when thrown in a method handler, will have their `message`
+ * serialized and sent across the wire. The sender will receive an equivalent error on the other side.
+ *
+ * Build-in types are included but developers may use any string, with a max length of 256 bytes.
+ */
+
+export class RpcError extends Error {
+  static MAX_MESSAGE_BYTES = 256;
+  static MAX_DATA_BYTES = 15360; // 15 KB
+
+  code: number;
+  data?: string;
+
+  /**
+   * Creates an error object with the given code and message, plus an optional data payload.
+   *
+   * If thrown in an RPC method handler, the error will be sent back to the caller.
+   *
+   * Error codes 1001-1999 are reserved for built-in errors (see RpcError.ErrorCode for their meanings).
+   */
+  constructor(code: number, message: string, data?: string) {
+    super(message);
+    this.code = code;
+    this.message = truncateBytes(message, RpcError.MAX_MESSAGE_BYTES);
+    this.data = data ? truncateBytes(data, RpcError.MAX_DATA_BYTES) : undefined;
+  }
+
+  toJSON() {
+    return {
+      code: this.code,
+      message: this.message,
+      data: this.data,
+    };
+  }
+
+  static ErrorCode = {
```

Comment:
can we device some sort of number based grouping for those already?
would e.g. be nice to have RECIPIENT_DISCONNECTED and RECIPIENT_NOT_FOUND in the same group.

Reply:
yeah sounds good, I'll group them thematically into 10xx, 11xx, 12xx etc in the final version.

Reply:
I went with 14xx and 15xx so they have some vague association to http